Rewrite .get_snmp_information()

bewing opened the pull request:
For napalm-automation/napalm-eos#113

<!-- Make sure you have read http://napalm.readthedocs.io/en/latest/contributing/index.html --!>

bewing committed:
Rewrite .get_snmp_information()

For napalm-automation/napalm-eos#113

bewing committed:
Update mocked data

bewing committed:
Remove extra comma from JSON file

bewing commented:
I'll fix the tests tomorrow morning, gotta run for now.

mirceaulinic commented on test/unit/mocked_data/test_get_snmp_information/normal/show_snmp_contact.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "contact": "noc@cloudflare.com"
```
On my device this is returned something like:

```
"contact": "\"noc@cloudflare.com\""
```

Which makes the output of this method:

```
u'contact': u'"noc@cloudflare.com"'
```

You only need to a strip those quotes when dumping the output - leaving inline comment where you need to apply changes.

bewing replied:
> On my device this is returned something like:
> > "contact": "\"noc@cloudflare.com\""



What version are you running?  I can't reproduce this behavior:

```
veos1(config)#snmp-server contact noc@cloudflare.com
veos1(config)#end
veos1#show snmp contact | json
{
    "contact": "noc@cloudflare.com"
}
veos1#show version | json
{
    "modelName": "vEOS",
    "internalVersion": "4.16.6M-3205780.4166M",
    "systemMacAddress": "00:15:5d:1b:db:cf",
    "serialNumber": "",
    "memTotal": 1897592,
    "bootupTimestamp": 1480670690.0,
    "memFree": 135396,
    "version": "4.16.6M",
    "architecture": "i386",
    "isIntlVersion": false,
    "internalBuildId": "e796e94c-ba3b-4355-afcf-ef0abfbfaee3",
    "hardwareRevision": ""
}
```

mirceaulinic replied:
Interesting.. I'm running 4.17.1.1F

mirceaulinic commented on napalm_eos/eos.py:
```diff
@@ -1090,35 +1093,43 @@ def get_route_to(self, destination='', protocol=''):
+                if k == 'chassisId':
+                    snmp_dict['chassis_id'] = v
+                else:
+                    snmp_dict[k] = v
```
```python
snmp_dict[k] = v.strip('"')
```

bewing replied:
> However, I don't like that we're doing now four requests to the device in order to solve a small problem.

run_commands is a single request in pyeapi, no matter how many commands are passed. So it's technically only two calls instead of one.  I'm trying to stick with JSON wherever possible to cut down on the amount of screen scraping that has to happen -- "show snmp community" is unfortunately still an unconverted command.  I've informed our account rep that we're interested in seeing the command converted sooner rather than later, which would put us back down to one request instead of two.

In terms of actual performance, I do not have any experience with in-depth Python profiling, but on an example run using "time", the execution difference is very small, and usually outweighed by network delays:

```
[bewing@host:~/napalm-eos] [napalm-eos] develop* ± time python test.py
{u'chassis_id': u'',
 u'community': {u'with-acl': {u'acl': u'ipv4-std-acl', u'mode': u'ro'}},
 u'contact': u'Brandon Ewing',
 u'location': u''}

real    0m0.610s
user    0m0.115s
sys     0m0.014s
[bewing@chcss015:~/napalm-eos] [napalm-eos] develop* ± git checkout iss113
Switched to branch 'iss113'
Your branch is up-to-date with 'origin/iss113'.
[bewing@host:~/napalm-eos] [napalm-eos] iss113* ± time python test.py
{u'chassis_id': u'',
 u'community': {u'blah': {u'acl': u'', u'mode': u'ro'},
                u'public': {u'acl': u'', u'mode': u'ro'},
                u'with-acl': {u'acl': u'ipv4-std-acl', u'mode': u'ro'}},
 u'contact': u'Brandon Ewing',
 u'location': u''}

real    0m0.575s
user    0m0.114s
sys     0m0.016s
```

mirceaulinic replied:
Yes, I agree.

Just please strip the `"` at the beginning and at the end  - if double qoutes happen to not be there, won't hurt.

mirceaulinic reviewed (changes requested):
Tested and works fine.

However, I don't like that we're doing now four requests to the device in order to solve a small problem.
Feel free to refactor the current behaviour, I am not a big TextFSM fan either. But at least, please don't decrease the performances: it really felt slower than before (no wonder when the number of requests increased by 300%).
Also, think about the following: if we'd introduce few new fields in the returned output, say `source_interface`, `host`, `traps` etc. what do you do then? Do you make 3/4/5 requests more?
I would recommend you to use: `show running-config | section snmp-server` to retrieve the whole SNMP config and from there select what you return.

ogenstad commented:
I created an example in #115 which would solve this. It's really just a very small change in the TextFSM template. Also in general for these kind of changes I think it would generally be safer to keep the old test and reference data as is and just create a new test so it's easy to see that nothing old breaks.

mirceaulinic commented:
Yes, I agree with @ogenstad - and I find I do prefer his proposal instead.
I will listen also what @dbarrosop thinks about that before taking any action.

dbarrosop commented:
> Also in general for these kind of changes I think it would generally be safer to keep the old test

It depends. There are a few cases to consider:

1. Internal commands/behavior changes. As the commands changes the test fail if the mocked data is not updated accordingly. So we have two cases:
   1.a. Default test case, if it's too complex and it's not testing any particular scenario, it's fine to completely rewrite the normal case.
   1.b. Extra cases for weird scenarios. These have to be addressed properly updating the mocked data to avoid reintroducing old bugs.
2. Internal commands/behavior doesn't change. Unless a problem is detected on the data, it is forbidden to touch those. A new one can be added and I really encourage people to add test cases triggering the bug they are fixing but old tests shouldn't be modified.

ogenstad commented:
@dbarrosop, it always depends. :)

bewing committed:
Fix return type and update unit test

Testing in vEOS 4.16.6M shows that "rw|ro" is ALWAYS present in the
config, even when omitted on entry.  Added more unit tests to cover
more possible community string formats according to EOS
configuration guide

mirceaulinic reviewed (approved):
This is approved.

I will let @dbarrosop to merge either this one or https://github.com/napalm-automation/napalm-eos/pull/115 - both are fine by me.
